### ISSUE TITLE
Questions on behavior of a few kinds of P4 tables in BMv2

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/behavioral-model#698]
I haven't dug through the behavioral-model code in enough detail to fully answer these questions yet, but I have some guesses that I was hoping could be confirmed, or pointed out how they are wrong.

It appears that the match_type of a table as a whole in BMv2 is "the maximum of the match_type's of all of its key fields", where: `range > ternary > lpm > exact`
reference: https://github.com/p4lang/p4c/blob/master/backends/bmv2/common/control.cpp#L310-L327

I am curious what BMv2's implementation of a table that has one ternary and one lpm field, as an example.  What I _think_ it does is that the lpm field is basically treated as a second ternary field, but its mask is restricted to be a prefix mask.  If there are multiple matching entries, the length of the prefixes of the lpm field is _ignored_, and only the control-plane configured priority of the table entry is used to determine which of possibly several matching entries is "the winner".

I believe the same is true for an lpm field when there is at least one range field in the table's key.

In other words, the "l" in "lpm" is perhaps a bit of a misnomer when an lpm field is in a table, where it isn't the maximum match_type, because the length of the prefixes has no effect on the determination of which of multiple matching entries is the winner.

But, given the description of behaviors above (if they are correct), it is supported to have an lpm field in a table that also has ternary or range fields.

**Comment by antoninbas** [comment]
We use the same names for individual match types and the bmv2 table "match type", when they are actually fairly different. The match type of a given expression in the match key determines the data plane matching semantics, as well as the control plane interface to the table. The bmv2 table "match type" is really an implementation choice. The most suitable implementation should be chosen after inspecting the different individual match types. This choice is being made by the compiler and included in the JSON (I believe for historical reasons), but really there should be nothing preventing bmv2 to make that decision itself. I haven't tried myself, but I think by editing the table "match type" in the JSON, one could force bmv2 to use a "ternary" implementation for a table with a single match with type exact.

I will now use <MATCH_TYPE> when talking about an individual field match type and _<match_type>_ for the bmv2-specific implementation.

The implementations are as follows:
 * _exact_: standard library unordered map
 * _lpm_: a custom trie
 * _ternary_: a linked list overlayed on a vector, with an unordered map cache
 * _range_: a linked list overlayed on a vector, with an unordered map cache

The rules to pick the optimal implementation for a given P4 table are:
 * if more than one LPM field, reject
 * if one RANGE field, use _range_
 * if one TERNARY field, use _ternary_
 * if one LPM field, use _lpm_
 * otherwise (only EXACT), use _exact_

So exactly what you described...

Your example is also correct. When a table has one LPM match and one TERNARY match, the _ternary_ implementation is picked. The control plane API still requires a `value/prefix` for the LPM match. The prefix is converted to a mask. It does *not* determine the priority of the _ternary_ entry, the rationale being that if a TERNARY field is included in the table, the user wants to be able to choose an arbitrary priority value.

**Comment by antoninbas** [comment]
I don't have an issue with you making the changes required to support multiple LPM matches in ternary tables. I would be surprised you needed to do more than just remove the check in the JSON parsing code: https://github.com/p4lang/behavioral-model/blob/master/src/bm_sim/P4Objects.cpp#L1654, although we'll need a gtest unit test to make sure it works.

Maybe at some point I was toying with the idea of using the prefix length as the priority even for ternary tables...

**Comment by jafingerhut** [comment]
We should talk more before you make a change that makes the prefix length the priority for ternary tables.  The main issue I see with that idea is that in a table with a ternary bit<10> field and an lpm field, there are `3^10` possible keys that have different value/masks for the ternary field, that could all have the _same_ prefix for the lpm field.  Many of those keys could match the same packet, and if they are all restricted to have the same priority as each other (the prefix length), the P4Runtime spec (for good reason, in my opinion) says that it is not defined which of the equal-priority entries will match.

I think I will start with the change to the documentation of the existing behavior, and include a footnote suggesting that perhaps p4c and/or behavioral-model could be generalized to allow multiple lpm fields in a table that has at least one ternary or range field.

**Comment by jafingerhut** [comment]
Created PR that only has additions to documentation: https://github.com/p4lang/behavioral-model/pull/699

**Comment by antoninbas** [comment]
Sorry I wasn't clear. I wasn't suggesting inferring the priority from the prefix length for ternary tables which include a LPM match. I meant that at some point in the past I may have been toying with the idea of doing so, which would explain the current behavior of rejecting ternary tables with multiple LPM matches.

**Comment by jafingerhut** [comment]
OK, I went back and re-read your earlier message, and see now how it can be understood in two ways, and I got the meaning you did not intend.  All good, then.

**Comment by jafingerhut** [comment]
Closing this issue, as the PR documenting the match_kind behavior for simple_switch has been merged.
